File: android/src/main/kotlin/com/example/native_home_widgets/data/WidgetDataStore.kt
package com.example.native_home_widgets.data

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map

/// DataStore-backed persistence for widget data.
///
/// Keys are optionally scoped by widgetId, allowing multiple widgets
/// to maintain independent data stores within the same DataStore file.
class WidgetDataStore(private val context: Context) {

    private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(
        name = "native_home_widgets"
    )

    /// Saves a key-value pair, optionally scoped to a specific widget.
    suspend fun save(key: String, value: String, widgetId: String? = null) {
        val dataKey = buildKey(key, widgetId)
        context.dataStore.edit { prefs ->
            prefs[dataKey] = value
        }
    }

    /// Reads a value by key, returns [defaultValue] if not found.
    suspend fun get(key: String, widgetId: String? = null, defaultValue: String? = null): String? {
        val dataKey = buildKey(key, widgetId)
        return context.dataStore.data.map { prefs ->
            prefs[dataKey] ?: defaultValue
        }.first()
    }

    /// Removes a specific key.
    suspend fun remove(key: String, widgetId: String? = null) {
        val dataKey = buildKey(key, widgetId)
        context.dataStore.edit { prefs ->
            prefs.remove(dataKey)
        }
    }

    /// Clears all data for a widget, or all data if [widgetId] is null.
    suspend fun clear(widgetId: String? = null) {
        context.dataStore.edit { prefs ->
            if (widgetId == null) {
                prefs.clear()
            } else {
                val prefix = "$widgetId:"
                val keysToRemove = prefs.asMap().keys.filter { it.name.startsWith(prefix) }
                keysToRemove.forEach { prefs.remove(it) }
            }
        }
    }

    /// Returns all stored entries as a map (for debugging / export).
    suspend fun getAll(): Map<String, String> {
        return context.dataStore.data.map { prefs ->
            prefs.asMap().mapKeys { it.key.name }
        }.first()
    }

    private fun buildKey(key: String, widgetId: String?): Preferences.Key<String> {
        val name = if (widgetId != null) "$widgetId:$key" else "global:$key"
        return stringPreferencesKey(name)
    }
}

File: ios/Classes/data/WidgetDataStore.swift
import Foundation

/// Persistence layer for widget data on iOS.
///
/// Uses `UserDefaults` with an App Group suite to share data between
/// the host app and widget extensions. Falls back to standard
/// `UserDefaults` if no App Group is configured.
class WidgetDataStore {
    private let defaults: UserDefaults

    /// Initializes the data store with an optional App Group identifier.
    ///
    /// - Parameter appGroupId: The App Group suite name. If nil or empty,
    ///   falls back to `UserDefaults.standard`.
    init(appGroupId: String? = nil) {
        if let groupId = appGroupId, !groupId.isEmpty {
            self.defaults = UserDefaults(suiteName: groupId) ?? .standard
        } else {
            self.defaults = UserDefaults(suiteName: NativeHomeWidgetsConstants.defaultAppGroupId) ?? .standard
        }
    }

    /// Saves a value for the given key, optionally scoped to a widget.
    func save(key: String, value: Any, widgetId: String? = nil) {
        let namespacedKey = buildKey(key: key, widgetId: widgetId)
        defaults.set(value, forKey: namespacedKey)
    }

    /// Reads a value by key, returning `defaultValue` if not found.
    func get(key: String, widgetId: String? = nil, defaultValue: Any? = nil) -> Any? {
        let namespacedKey = buildKey(key: key, widgetId: widgetId)
        return defaults.object(forKey: namespacedKey) ?? defaultValue
    }

    /// Removes a specific key.
    func remove(key: String, widgetId: String? = nil) {
        let namespacedKey = buildKey(key: key, widgetId: widgetId)
        defaults.removeObject(forKey: namespacedKey)
    }

    /// Clears all data for a widget, or all data if `widgetId` is nil.
    func clear(widgetId: String? = nil) {
        let allKeys = defaults.dictionaryRepresentation().keys
        if let widgetId = widgetId {
            let prefix = "\(widgetId):"
            allKeys.filter { $0.hasPrefix(prefix) }.forEach {
                defaults.removeObject(forKey: $0)
            }
        } else {
            let prefix = "global:"
            allKeys.filter { $0.hasPrefix(prefix) || $0.contains(":") }.forEach {
                defaults.removeObject(forKey: $0)
            }
        }
    }

    /// Returns all stored entries as a dictionary.
    func getAll() -> [String: Any] {
        return defaults.dictionaryRepresentation()
    }

    private func buildKey(key: String, widgetId: String?) -> String {
        if let widgetId = widgetId {
            return "\(widgetId):\(key)"
        }
        return "global:\(key)"
    }
}
